interop/tapir/src/main/scala/caliban/interop/tapir/HttpUploadAdapter.scala

```diff
@@ -0,0 +1,120 @@
+                             case _: DecodeResult.Failure => ZIO.fail(TapirResponse(StatusCode.BadRequest))
+                             case DecodeResult.Value(v)   => ZIO.succeed(v)
+                           }
+          rawMap        <- ZIO.fromOption(partsMap.get("map")) orElseFail TapirResponse(StatusCode.BadRequest)
+          map           <- mapCodec.rawDecode(new String(rawMap.body, StandardCharsets.UTF_8)) match {
+                             case _: DecodeResult.Failure => ZIO.fail(TapirResponse(StatusCode.BadRequest))
+                             case DecodeResult.Value(v)   => ZIO.succeed(v)
+                           }
+          filePaths      = map.map { case (key, value) => (key, value.map(parsePath).toList) }.toList
+                             .flatMap(kv => kv._2.map(kv._1 -> _))
+          handler        = Uploads.handler(handle =>
+                             ZIO
+                               .succeed(partsMap.get(handle))
+                               .some
+                               .flatMap(fp =>
+                                 Random.nextUUID.asSomeError
+                                   .map(uuid =>
+                                     FileMeta(
+                                       uuid.toString,
+                                       fp.body,
+                                       fp.contentType,
+                                       fp.fileName.getOrElse(""),
+                                       fp.body.length
+                                     )
+                                   )
+                               )
+                               .unsome
+                           )
+          uploadQuery    = GraphQLUploadRequest(request, filePaths, handler)
+          query          = serverRequest.headers
+                             .find(isFtv1Header)
+                             .fold(uploadQuery.remap)(_ => uploadQuery.remap.withFederatedTracing)
+          response      <- executeRequest(query, serverRequest)
+                             .provideSomeLayer[R](ZLayer(uploadQuery.fileHandle))
+        } yield response
+
+      io.either
+    }
+
+    endpoint.serverLogic(logic)
+  }
+
+  protected def executeRequest(
+    graphQLRequest: GraphQLRequest,
+    serverRequest: ServerRequest
+  ): ZIO[R, TapirResponse, GraphQLResponse[E]]
+
+  def configure[R1](configurator: ZLayer[R1 & ServerRequest, TapirResponse, R]): HttpUploadAdapter[R1, E] =
+    new HttpUploadAdapter[R1, E] {
+      val endpoint: PublicEndpoint[UploadRequest, TapirResponse, GraphQLResponse[E], Any] =
+        self.endpoint
+
+      def executeRequest(
+        graphQLRequest: GraphQLRequest,
+        serverRequest: ServerRequest
+      ): ZIO[R1, TapirResponse, GraphQLResponse[E]] =
+        self
+          .executeRequest(graphQLRequest, serverRequest)
+          .provideSome[R1](ZLayer.succeed(serverRequest), configurator)
+    }
+}
```

Comment:
Just a thought, but if we made this trait sealed could we then optimize the `provideSome` by doing something like:

```scala
case class Configured[R1, R](layer: ZLayer[R1 & ServerRequest, TapirResponse, R]) extends HttpUploadAdapter[R1, E] {
  def configure[R2](configurator: ZLayer[R2 & ServerRequest, TapirResponse, R1]) = 
    Configured(ZLayer.make[R1](configurator, layer))
}
```

I haven't tried to compile that so not sure if that would work.

Reply:
Applied on `HttpAdapter`. Let me know what you think, then I'll apply elsewhere.